kotlinx-coroutines-core/common/src/internal/SegmentList.kt

```diff
@@ -0,0 +1,213 @@
+ * (e.g., queues should use it in dequeue operations).
+ *
+ * Since segments can be removed from the list, or it can be closed for further segment additions, this function
+ * returns the segment `s` with `s.id >= id` or `CLOSED` if all the segments in this linked list have lower `id`
+ * and the list is closed.
+ */
+internal inline fun <S : Segment<S>> AtomicRef<S>.findSegmentAndMoveForward(id: Long, startFrom: S, createNewSegment: (id: Long, prev: S?) -> S): SegmentOrClosed<S> {
+    while (true) {
+        val s = startFrom.findSegmentInternal(id, createNewSegment)
+        if (s.isClosed || moveForward(s.segment)) return s
+    }
+}
+
+/**
+ * Closes this linked list of segments by forbidding adding new segments,
+ * returns the last segment in the list.
+ */
+internal fun <S : Segment<S>> S.close(): S {
+    var cur: S = this
+    while (true) {
+        val next = cur.nextOrClosed.run { if (isClosed) return cur else segment }
+        if (next === null) {
+            if (cur.markAsClosed()) return cur
+        } else {
+            cur = next
+        }
+    }
+}
+
+/**
+ * Each segment in [SegmentList] has a unique id and is created by [SegmentList.newSegment].
+ * Essentially, this is a node in the Michael-Scott queue algorithm, but with
+ * maintaining [prev] pointer for efficient [remove] implementation.
+ */
+internal abstract class Segment<S : Segment<S>>(val id: Long, prev: S?, pointers: Int) {
+    // Pointer to the next segment, updates similarly to the Michael-Scott queue algorithm.
+    private val _next = atomic<Any?>(null)
+    val nextOrClosed: NextSegmentOrClosed<S> get() = NextSegmentOrClosed(_next.value)
+    fun trySetNext(value: S): Boolean = _next.compareAndSet(null, value)
+
+    // Pointer to the previous segment, updates in [remove] function.
+    val prev = atomic(prev)
+
+    /**
+     * Cleans the pointer to the previous segment.
+     */
+    fun cleanPrev() = prev.lazySet(null)
+
+    /**
+     * This property should return the maximal number of slots in this segment,
+     * it is used to define whether the segment is logically removed.
+     */
+    abstract val maxSlots: Int
+
+    // numbers of cleaned slots (lowest bits) and AtomicRef pointers to this segment (highest bits)
+    private val cleanedAndPointers = atomic(pointers shl POINTERS_SHIFT)
+
+    /**
+     * Returns `true` if this segment is logically removed from the queue.
+     * The segment is considered as removed if all the slots are cleaned,
+     * there is no pointers to this segment from outside, and
+     * it is not a physical tail in the linked list of segments.
+     */
+    inline val removed get() = cleanedAndPointers.value == maxSlots && _next.value !== null
+
+    // increments the number of pointers if this segment is not logically removed\
+    inline fun tryIncPointers() = cleanedAndPointers.addConditionally(1 shl POINTERS_SHIFT) { it != maxSlots }
```

Comment:
The segment may still be present (not logically removed) when this method returns `false` if it's the tail segment.

This may be a significant problem. Assume we have segments 1 and 2, with `tail` pointing to `segment_1`, and all the cells in 2 are cancelled (as with the comment for `moveForward`, this can't ever happen in the semaphore implementation, for the same reasons). If we call `tail.findSegmentAndMoveForward(2, segment_1, ...)`, `findSegmentInternal` returns `segment_2`. It isn't closed, and `moveForward` fails because the call `tryIncPointers` fails, so we loop. The situation hasn't changed, so we may loop forever.

Reply:
Changed the condition: `{ it != maxSlots || _next.value == null }`